Core/Src/main.c

```diff
@@ -1,181 +1,223 @@
+  */
+/* USER CODE END Header */
+/* Includes ------------------------------------------------------------------*/
+#include "main.h"
+#include "adc.h"
+#include "spi.h"
+#include "tim.h"
+#include "usart.h"
+#include "gpio.h"
+
+/* Private includes ----------------------------------------------------------*/
+/* USER CODE BEGIN Includes */
+
+/* USER CODE END Includes */
+
+/* Private typedef -----------------------------------------------------------*/
+/* USER CODE BEGIN PTD */
+
+/* USER CODE END PTD */
+
+/* Private define ------------------------------------------------------------*/
+/* USER CODE BEGIN PD */
+/* USER CODE END PD */
+
+/* Private macro -------------------------------------------------------------*/
+/* USER CODE BEGIN PM */
+
+/* USER CODE END PM */
+
+/* Private variables ---------------------------------------------------------*/
+
+/* USER CODE BEGIN PV */
+HAL_StatusTypeDef hal_status;
+uint8_t data_size = 3;
+uint32_t timeout = 1 << 31;
+
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+/* USER CODE BEGIN PFP */
+
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+
+/* USER CODE END 0 */
+
+/**
+  * @brief  The application entry point.
+  * @retval int
+  */
+int main(void)
+{
+  /* USER CODE BEGIN 1 */
+
+  /* USER CODE END 1 */
+
+  /* MCU Configuration--------------------------------------------------------*/
+
+  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
+  HAL_Init();
+
+  /* USER CODE BEGIN Init */
+
+  /* USER CODE END Init */
+
+  /* Configure the system clock */
+  SystemClock_Config();
+
+  /* USER CODE BEGIN SysInit */
+
+  /* USER CODE END SysInit */
+
+  /* Initialize all configured peripherals */
+  MX_GPIO_Init();
+  MX_USART2_UART_Init();
+  MX_SPI1_Init();
+  MX_TIM1_Init();
+  MX_ADC_Init();
+  /* USER CODE BEGIN 2 */
+
+  HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_SET);
+
+  uint8_t tx_buffer[3];
+  tx_buffer[0] = 0x01;
+  tx_buffer[1] = 0x80;
+  uint8_t rx_buffer[3];
+
+  //pull CS pin low.
+  HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_RESET);
+
+  // Start timer
+  HAL_TIM_PWM_Start(&htim1, TIM_CHANNEL_1);
+
+  /* USER CODE END 2 */
+
+  /* Infinite loop */
+  /* USER CODE BEGIN WHILE */
+  while (1)
+  {
+	  hal_status =  HAL_SPI_TransmitReceive(&hspi1, tx_buffer, rx_buffer, data_size, timeout); // transmit/receive data to/from ADC.
+
+	  uint8_t second_byte =  rx_buffer[1];
+	  uint8_t third_byte = rx_buffer[2];
+
+	  // We want the last 2 bits of the second received byte and the full byte of the third received byte.
+	  uint16_t analog_voltage_val = ((second_byte & 0x03) << 8) + third_byte;
+
+	  // map ADC value to a value between 480 and 960 for 5 - 10% duty cycle since counter period is 9600.
+	  int mapped_val = (480.0/1024) * analog_voltage_val;

```

Comment:
so in this case, although I may know what these numbers mean, many other people may not. 480 and 1024 are considered magic numbers, we should be setting them to some constant or macro.

Also, maybe a uint would better describe the value bec the value will never be negative.